python/processing/algfactory.py

```diff
@@ -0,0 +1,396 @@
+                       QgsProcessingOutputHtml,
+                       QgsProcessingOutputLayerDefinition,
+                       QgsProcessingOutputMapLayer,
+                       QgsProcessingOutputMultipleLayers,
+                       QgsProcessingOutputNumber,
+                       QgsProcessingOutputRasterLayer,
+                       QgsProcessingOutputVectorLayer)
+
+from qgis.core import QgsMessageLog
+
+
+def print(*args, **kw):
+    QgsMessageLog.logMessage(" ".join(map(str, args)), "Factory")
+
+
+def make_output(**args):
+    cls = args['cls']
+    del args['cls']
+    newargs = {
+        "name": args['name'],
+        "description": args['description'],
+    }
+    return cls(**newargs)
+
+
+class ProcessingAlgFactoryException(Exception):
+    def __init__(self, message):
+        super(ProcessingAlgFactoryException, self).__init__(message)
+
+
+class AlgWarpper(QgsProcessingAlgorithm):
+
+    def __init__(self, name=None, display=None,
+                 group=None, group_id=None, inputs=None,
+                 outputs=None, func=None, help=None, icon=None):
+        super(AlgWarpper, self).__init__()
+        self._inputs = OrderedDict(inputs or {})
+        self._outputs = OrderedDict(outputs or {})
+        self._icon = icon
+        self._name = name
+        self._group = group
+        self._group_id = group_id
+        self._display = display
+        self._func = func
+        self._help = help
+        ## Only take the help from the function if it hasn't already been set.
+        if self._func and not self._help:
+            self._help = self._func.__doc__
+
+    def _get_parent_id(self, parent):
+        if isinstance(parent, str):
+            return parent
+        else:
+            raise NotImplementedError()
+
+    # Wrapper logic
+    def define(self, name, label, group, group_label, help=None, icon=None):
+        self._name = name
+        self._display = label
+        self._group = group_label
+        self._group_id = group
+        self._help = help
+        self._icon = icon
+
+    def end(self):
+        if not self.has_inputs:
+            raise ProcessingAlgFactoryException("No inputs defined for '{}' alg. "
+                                                "At least one is required. Use @alg.input"
+                                                "to define one.".format(self.name()))
+        if not self.has_outputs:
+            raise ProcessingAlgFactoryException("No outputs defined for '{}' alg"
+                                                "At least one must be defined. Use @alg.output")
+
+    def add_output(self, type, **kwargs):
+        parm = self._add_parm(type, output=True, **kwargs)
+        self._outputs[parm.name()] = parm
+
+    def add_help(self, helpstring, *args, **kwargs):
+        self._help = helpstring
+
+    def add_input(self, type, **kwargs):
+        parm = self._add_parm(type, **kwargs)
+        self._inputs[parm.name()] = parm
```

Comment:
Worth checking if an input with this name already exists?

Reply:
@m-kuhn already does inside `_add_parm` but I will refactor this and make it more clear.